### ISSUE TITLE
Configuration defaults passed via workbench options don't work

### ISSUE THREAD

**Issue by connor4312** [microsoft/vscode#224236]
1. Register some [configuration defaults](https://github.com/microsoft/vscode/blob/25dd3cf86fb66e1e3dab023458b3cf2d0d55c69c/src/vs/workbench/browser/web.api.ts#L299) using the web API (e.g. [here](https://github.com/microsoft/vscode/blob/25dd3cf86fb66e1e3dab023458b3cf2d0d55c69c/src/vs/code/browser/workbench/workbench.ts#L578) for the code-server.sh script)
2. 🐛 notice they are still undefined when read from the local extension host (at least before a remote is resolved)

It _seems_ like it should work because the overrides are registered here

https://github.com/microsoft/vscode/blob/25dd3cf86fb66e1e3dab023458b3cf2d0d55c69c/src/vs/workbench/services/configuration/browser/configuration.ts#L48-L50

before the MainThreadConfiguration is even instantiated, so I would expect they get passed during initial hydration here

https://github.com/microsoft/vscode/blob/25dd3cf86fb66e1e3dab023458b3cf2d0d55c69c/src/vs/workbench/api/browser/mainThreadConfiguration.ts#L29

...but they do not. I did some debugging myself but got a bit lost in config land.

Having these options is needed for a couple partner team feature asks.

**Comment by sandy081** [comment]
Assigning to August for investigation.

**Comment by sandy081** [comment]
I changed the default of `workbench.colorTheme` setting and it works (github.dev also uses the same). You should have the setting registered for the default to be applied. I suspect the setting is not registered in your scenario?

**Comment by connor4312** [comment]
The setting is registered. Repro is a little annoying since it requires an extension doing the resolution

1. Run vscode.dev locally so that you can debug things (`npm install && npm run build && node out/server/main.js`)
2. Clone https://github.com/microsoft/vscode-remote-tunnels. `git checkout connor4312/preference-auth-provider && npm install && npm run watch` in one terminal, and `cd extension && npx serve -p 3001 --cors` in another
3. With devtools open, hit http://localhost:3000/+aHR0cDovL2xvY2FsaG9zdDozMDAx/some-tunnel-here (first segment is base64-encoded localhost:3001; the tunnel does not actually need to be hosted.) You can add `?vscode-quality=dev` to that URL to use your local vscode build found in a sibling folder to vscode-dev.
4. Notice that even though we set a default value for [remote.tunnels.useAuthProvider](https://github.com/microsoft/vscode-remote-tunnels/blob/c4e3fd551e402661685bec41949bc0f57630294a/extension/src/webOpener/webOpener.ts#L57), it's undefined when you hit the `debugger` statement.

**Comment by sandy081** [comment]
Thanks for the steps. I can confirm that the setting is not yet registered at the time when you are accessing this setting in your extension. Your extension (remote resolver) is accessing this setting while resolving the authority and I suspect, contribution points of this extension are not read/handled yet.

@alexdima Can you please confirm if this is true.

**Comment by alexdima** [comment]
Yes, this is an old issue since we first did remotes https://github.com/microsoft/vscode/issues/88044 . Extension points are not handled until after the resolver resolves. In code:
  * `AbstractExtensionService._initialize` first waits for `_resolveExtensions`
  * `NativeExtensionService._resolveExtensions` does the remote authority resolving
  * Only then is `_processExtensions` invoked
  * That one eventually calls `_doHandleExtensionPoints` which goes through our extension points to inform them of extension contributions

There are some historical reasons behind it, I think back then we did not support disabling extensions without a window reload, so the extension points themselves were not ready for an extension point removal. It also did not seem to be so bad of an issue because the only affected extensions would be our resolver extensions and they mostly suffered when reading their own default configs, which they should normally know since they contribute their extensions. But it looks like you have a new pattern where an embedder is driving the resolver's default config value, which makes the initial issue more severe.

I don't know if a workaround can be made in the configuration service to be able to give values for settings that don't have their schema registered @sandy081  or if we now must handle at least the resolver's extension points before resolving. If it's the latter, then @connor4312 maybe you can look into a PR. I think something must be added before invoking the resolver [here](https://github.com/microsoft/vscode/blob/f221c545c23d8298dde710012084eadbb67b8e2c/src/vs/workbench/services/extensions/electron-sandbox/nativeExtensionService.ts#L361-L376) which would potentially call `_doHandleExtensionPoints`. But be aware that it would not be good to handle the extension points of all locally available extensions since many of them will be available remotely and that would just cause a lot of churn. Now that I write this, I remember the second reason why I didn't tackle this. Some extension point handlers use the fact that they are being called as a way to think of themselves as being "ready"/"finished", which in this situation is clearly not the case. I think the debugger was one of them, but maybe others, I don't remember.

**Comment by sandy081** [comment]
> if a workaround can be made in the configuration service to be able to give values for settings that don't have their schema registered

I feel this allows misusing of the API for any random property. 

How about handling only `configuration` contribution of a resolver extension? Or as a generic approach, let the contribution point define if it is ready for handling a contribution point during resolving phase?

**Comment by connor4312** [comment]
> How about handling only configuration contribution of a resolver extension? Or as a generic approach, let the contribution point define if it is ready for handling a contribution point during resolving phase?

Sure, I think either of these would work.

**Comment by connor4312** [comment]
This remains an important scenario for our friends at Azure Functions, is there any way we can figure a solution here?

**Comment by sandy081** [comment]
I will look into it this milestone/

**Comment by connor4312** [comment]
Thanks so much!

**Comment by sandy081** [comment]
I would request @connor4312 to verify this as he has the set up and can also check if this is working for his scenario.

**Comment by connor4312** [comment]
Yep it works thanks!
